enterprise/suppress-user/internal/badgerdb/badgerdb.go

```diff
@@ -0,0 +1,307 @@
+package badgerdb
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/dgraph-io/badger/v3"
+	"github.com/dgraph-io/badger/v3/options"
+	"github.com/rudderlabs/rudder-server/enterprise/suppress-user/model"
+	"github.com/rudderlabs/rudder-server/utils/logger"
+	"github.com/samber/lo"
+)
+
+// the key used in badgerdb to store the current token
+const tokenKey = "__token__"
+
+// Opt is a function that configures a badgerdb repository
+type Opt func(*Repository)
+
+// WithSeederSource sets the source of the seed data
+func WithSeederSource(seederSource func() (io.Reader, error)) Opt {
+	return func(r *Repository) {
+		r.seederSource = seederSource
+	}
+}
+
+// WithMaxSeedWait sets the maximum time to wait for the seed to complete.
+// If the seed takes longer than this, the repository will be started in restoring state and all
+// repository methods will return [ErrRestoring] until the seed completes. The default wait time is 10 seconds.
+func WithMaxSeedWait(maxSeedWait time.Duration) Opt {
+	return func(r *Repository) {
+		r.maxSeedWait = maxSeedWait
+	}
+}
+
+// Repository is a repository backed by badgerdb
+type Repository struct {
+	// logger to use
+	log logger.Logger
+	// path to the badger db directory
+	path string
+	// max number of goroutines to use (badger config)
+	maxGoroutines int
+
+	maxSeedWait  time.Duration
+	seederSource func() (io.Reader, error)
+
+	db *badger.DB
+
+	// lock to prevent concurrent access to db during restore
+	restoringLock sync.RWMutex
+	restoring     bool
+	closeOnce     sync.Once
+	closed        chan struct{}
+}
+
+// NewRepository returns a new repository backed by badgerdb.
+func NewRepository(basePath string, log logger.Logger, opts ...Opt) (*Repository, error) {
+	b := &Repository{
+		log:           log,
+		path:          path.Join(basePath, "badgerdbv3"),
+		maxGoroutines: 1,
```

Comment:
since, `maxGoroutines` could be really helpful during `Restore()`, to decrease restore time. Shouldn't we should consider making it configurable and maybe increase the default value?

Reply:
According to benchmarks, `maxGoroutines` affects memory usage considerably and doesn't affect performance for our use case. This is the reason why it was set to 1.